File: flyweight/java/FlyweightDemo.java
import java.util.HashMap;
import java.util.Map;
import java.util.Random;

interface Shape {

	void draw();
}

final class Circle implements Shape {

	private int x, y, radius;
	private String color;

	public Circle(String color) {
		this.color = color;
	}

	public void setX(int x) {
		this.x = x;
	}

	public void setY(int y) {
		this.y = y;
	}

	public void setRadius(int radius) {
		this.radius = radius;
	}

	@Override
	public void draw() {
		System.out.printf("[Draw circle] x: %d, y: %d, radius: %d, color: %s\n",
			this.x, this.y, this.radius, this.color);
	}
}

final class CircleFactory {

	private static final Map<String, Shape> flyweights = new HashMap<>();

	private CircleFactory() {}

	public static Shape getCircle(String color) {
		Shape circle = flyweights.get(color);

		if (circle == null) {
			circle = new Circle(color);
			flyweights.put(color, circle);
		}

		return circle;
	}
}

final class FlyweightDemo {

	public static void main(String[] args) {
		String[] colors = {"Red", "Green", "Blue", "White", "Black"};
		Random ran = new Random();

		ran.ints(20, 0, colors.length).forEach(n -> {
			Circle circle = (Circle)CircleFactory.getCircle(colors[n]);
			circle.setX(ran.nextInt(100));
			circle.setY(ran.nextInt(100));
			circle.setRadius(100);
			circle.draw();
		});
	}
}

File: flyweight/kotlin/flyweight_demo.kt
import kotlin.random.*

interface Shape {

	fun draw(): Unit
}

class Circle(private val color: String) : Shape {

	var x: Int
	var y: Int
	var radius: Int

	init {
		this.x = 0
		this.y = 0
		this.radius = 0
	}

	override fun draw(): Unit {
		println("[Draw circle] x: ${this.x}, y: ${this.y}, radius: ${this.radius}, color: ${this.color}\n")
	}
}

class CircleFactory {

	companion object {
		private val flyweights: MutableMap<String, Shape> = mutableMapOf()

		fun getCircle(color: String): Shape =
			flyweights.getOrPut(color) { Circle(color) }
	}

	private constructor() {}
}

fun main(): Unit {
	val colors: Array<String> = arrayOf("Red", "Green", "Blue", "White", "Black")

	for(i in 1..20) {
		val circle: Circle = CircleFactory.getCircle(colors.random()) as Circle;
		circle.x = Random.nextInt(100)
		circle.y = Random.nextInt(100)
		circle.radius = 100
		circle.draw()
	}
}